File: Group D/Code/golang/BGR2GRAY.go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	"gocv.io/x/gocv"
)

func main() {
	startTime := gocv.GetTickCount()

	inputFolder := "../../opencv_dataset/Blue"

	outputFolder := "B2GDone"
	if err := os.MkdirAll(outputFolder, os.ModePerm); err != nil {
		fmt.Printf("Error creating output folder: %v\n", err)
		return
	}

	files, err := filepath.Glob(filepath.Join(inputFolder, "*.jpg"))
	if err != nil {
		fmt.Printf("Error reading input folder: %v\n", err)
		return
	}
	for _, filePath := range files {
		filename := filepath.Base(filePath)
		image := gocv.IMRead(filePath, gocv.IMReadColor)
		if image.Empty() {
			fmt.Printf("Error reading image: %s\n", filename)
			os.Exit(1)
		}
		grayImage := gocv.NewMat()
		gocv.CvtColor(image, &grayImage, gocv.ColorBGRToGray)

		outputPath := filepath.Join(outputFolder, filename)
		if ok := gocv.IMWrite(outputPath, grayImage); !ok {
			fmt.Printf("Error writing image: %s\n", filename)
		}
		fmt.Printf("Image converted, new image saved: %s\n", outputPath)
	}

	endTime := gocv.GetTickCount()
	elapsedTime := (endTime - startTime) / gocv.GetTickFrequency()
	fmt.Printf("Execution Time: %v seconds\n", elapsedTime)
}

File: Group D/Code/rust/BGR2GRAY.rs
use opencv::{imgcodecs, imgproc, core, types};
use std::fs;
use std::env;

#[derive(Debug)]
enum CustomError {
    OpenCVError(opencv::Error),
    IOError(std::io::Error),
}

impl From<opencv::Error> for CustomError {
    fn from(error: opencv::Error) -> Self {
        CustomError::OpenCVError(error)
    }
}

impl From<std::io::Error> for CustomError {
    fn from(error: std::io::Error) -> Self {
        CustomError::IOError(error)
    }
}

fn main() -> Result<(), CustomError> {
    if let Ok(cwd) = env::current_dir() {
        println!("Current working directory: {:?}", cwd);
    } else {
        eprintln!("Failed to get the current working directory.");
    }

    let input_folder = "../../../../opencv_dataset/Blue";
    let output_folder = "../../B2GDone";

    let entries = fs::read_dir(input_folder)?;
    let tick_frequency = core::get_tick_frequency()?;
    let start_time = core::get_tick_count()?;

    for entry in entries {
        let entry = entry?;
        let path = entry.path();

        if !path.is_file() {
            continue;
        }

        if let Some(ext) = path.extension() {
            if ext == "jpg" || ext == "jpeg" {
                if let Ok(mut image) = imgcodecs::imread(path.to_str().unwrap(), imgcodecs::IMREAD_COLOR) {
                    let mut gray_image = core::Mat::default();
                    imgproc::cvt_color(&mut image, &mut gray_image, imgproc::COLOR_BGR2GRAY, 0);

                    let file_name = path.file_name().unwrap().to_string_lossy();
                    let output_path = format!("{}/{}", output_folder, file_name);

                    imgcodecs::imwrite(&output_path, &gray_image, &types::VectorOfi32::new())?;
                    println!("Image converted, new image saved: {}", output_path);
                }
            }
        }
    }

    let end_time = core::get_tick_count()?;
    let execution_time = (end_time as f64 - start_time as f64) / tick_frequency;

    println!("Execution Time: {:.6} seconds", execution_time);

    Ok(())
}
